Python function vx_rpp implementation

SundarRajan28 opened the pull request.

SundarRajan28 committed:
Add external source operator node

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into external_source_operator

SundarRajan28 committed:
Merge branch 'ROCm:develop' into external_source_operator

SundarRajan28 committed:
Temp changes (NWC)

SundarRajan28 committed:
Add python function support

SundarRajan28 committed:
Fix issues with PythonFunction openvx node

SundarRajan28 committed:
Update PythonFunction node to run callback

SundarRajan28 committed:
Remove dlsym method for running callback

SundarRajan28 committed:
Remove ROI from openvx node

SundarRajan28 committed:
Add API docs

SundarRajan28 committed:
Merge branch 'ROCm:develop' into python_function_callback

shobana-mcw commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -2055,8 +2055,22 @@ extern "C"
+	/*! \brief [Graph] Calls the python function, passes the input as numpy array and returns the output.
+	 * \ingroup group_amd_rpp
+	 * \param [in] graph The handle to the graph.
+	 * \param [in] pSrc The array of input tensor in <tt>\ref VX_TYPE_UINT8</tt> or <tt>\ref VX_TYPE_FLOAT32</tt> or <tt>\ref VX_TYPE_FLOAT16</tt> or <tt>\ref VX_TYPE_INT8</tt> format data.
```
Check the description wrt input tensor type.

shobana-mcw commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -2055,8 +2055,22 @@ extern "C"
+	 * \param [out] pDst The output tensor in <tt>\ref VX_TYPE_UINT8</tt> or <tt>\ref VX_TYPE_FLOAT32</tt> or <tt>\ref VX_TYPE_FLOAT16</tt> or <tt>\ref VX_TYPE_INT8</tt> format data.
+	 * \param [in] bridgeFnPtr The pointer to the bridge function in <tt>\ref VX_TYPE_UINT64</tt> format.
+	 * \param [in] functionId The python function ID in <tt>\ref VX_TYPE_UINT64</tt> format.
+	 * \param [in] inputLayouts The input layouts in <tt>\ref VX_TYPE_ARRAY</tt> (item type <tt>\ref VX_TYPE_INT32</tt>) denoting the layout of each input tensor.
```
Do we nee input and output layouts?

SundarRajan28 replied:
We need the layout info for setting description pointers

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+
+// Local copy of rocAL Python bridge C ABI (keep in sync with rocAL)
+#ifndef ROCAL_PY_MAX_TENSOR_DIMS
+#define ROCAL_PY_MAX_TENSOR_DIMS 8
```
Why are we setting it to 8?

SundarRajan28 replied:
RPP_MAX_TENSOR_DIMS is set as 5 so I am setting it as 5

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+    uint32_t device_type;        /* AGO_TARGET_AFFINITY_{CPU,GPU}; currently CPU-only */
+} RocalPyExecParams;
+
+typedef vx_status (*rocal_process_python_function_fn)(void *src_ptr, void *dst_ptr, const RocalPyExecParams *params);
```
Add a comment here

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+struct PythonFunctionLocalData {
```
Check if all the member variables are needed.

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+    return status;
+}
+
+static vx_status VX_CALLBACK validatePythonFunction(vx_node /*node*/, const vx_reference parameters[], vx_uint32 /*num*/, vx_meta_format metas[]) {
```
Why node and num are commented?

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+    vx_enum scalar_type;
+    STATUS_ERROR_CHECK(vxQueryScalar((vx_scalar)parameters[2], VX_SCALAR_TYPE, &scalar_type, sizeof(scalar_type)));
+    if (scalar_type != VX_TYPE_UINT64)
+        return ERRMSG(VX_ERROR_INVALID_TYPE, "PythonFunction validate: Parameter #3 (bridgeFnPtr) must be UINT64\n");
```
Check the parameter number.

SundarRajan28 replied:
The parameter indexes starts from 0 so I add +1 to the index for the error message

shobana-mcw replied:
You are querying for parameter 2 but error message is printed for parameter 3. Please check for other parameters.

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+        return VX_ERROR_NOT_IMPLEMENTED;
+    }
+
+    RocalPyExecParams p{};
```
Rename the parameter.

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+#define ROCAL_PY_MAX_TENSOR_DIMS 8
+#endif
+
+typedef struct RocalPyTensorDesc_ {
```
Why there is a _? Please check coding guidelines.

shobana-mcw commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,317 @@
+        return VX_ERROR_NOT_IMPLEMENTED;
+
+    if (!data->bridge_fn) {
+        vxAddLogEntry((vx_reference)node, VX_ERROR_NOT_IMPLEMENTED, "PythonFunction bridge function not available. Build rocAL with Python bridge.\n");
```
I think the error message is a bit ambiguous.

shobana-mcw reviewed (commented):
Please address the review comments.

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into python_function_callback

SundarRajan28 committed:
Resolve review comments

SundarRajan28 committed:
Fix API comment

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -2055,8 +2055,22 @@ extern "C"
 	 */
 	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppLog1p(vx_graph graph, vx_tensor pSrc, vx_tensor pSrcRoi, vx_tensor pDst, vx_scalar inputLayout);
 
+	/*! \brief [Graph] Calls the python function, passes the input as numpy array and returns the output.
```
Maybe we need to rephrase the def saying

Invokes python function and returns output, accepts input as numpy array, Please check

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -2055,8 +2055,22 @@ extern "C"
+	/*! \brief [Graph] Calls the python function, passes the input as numpy array and returns the output.
+	 * \ingroup group_amd_rpp
+	 * \param [in] graph The handle to the graph.
+	 * \param [in] pSrc The input tensor in <tt>\ref VX_TYPE_UINT8</tt> or <tt>\ref VX_TYPE_FLOAT32</tt> or <tt>\ref VX_TYPE_FLOAT16</tt> or <tt>\ref VX_TYPE_INT8</tt> format data.
```
It can support int32 types as well right, include add supported types of pSrc and pDst

fiona-gladwin replied:
Is this done?

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,315 @@
+
+#include <stdint.h>
+
+#include <cstdio>
```
Are all these includes required please check

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,315 @@
+#define ROCAL_PY_MAX_TENSOR_DIMS 5
+#endif
+
+typedef struct {
```
Add a comment saying it mirrors the rocAL structs kind

SundarRajan28 replied:
Comment added in L31

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    vx_status status = VX_SUCCESS;
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HOST, &data->pSrc, sizeof(data->pSrc)));
```
In case the HIP is passed we need to raise an error code in this file, Please make that condition check somewhere

SundarRajan28 replied:
The HIP check is added in process function

fiona-gladwin commented on amd_openvx_extensions/amd_rpp/source/tensor/PythonFunction.cpp:
```diff
@@ -0,0 +1,315 @@
+    data->pDstGenericDesc->dataType = getRpptDataType(output_tensor_dtype);
+    data->pDstGenericDesc->offsetInBytes = 0;
+    fillGenericDescriptionPtrfromDims(data->pDstGenericDesc, data->outputLayout, data->outputTensorDims);
+
```
Remove one line

SundarRajan28 committed:
Resolve review comments

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into python_function_callback

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.26.0 to 1.27.0 in /docs/sphinx (#1572)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.26.0 to 1.27.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.26.0...v1.27.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-version: 1.27.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

SundarRajan28 committed:
Merge branch 'develop' into python_function_callback

SundarRajan28 committed:
Update vx_ext_rpp version

SundarRajan28 committed:
Fix CI build issue

SundarRajan28 committed:
Update changelog

SundarRajan28 committed:
Update error msg
